### PR TITLE
✨Feat(source-s3): add file-transfer feature to s3

### PR DESCRIPTION
**Pull request by aldogonzalez8**
## What
Implement file-transfer feature for source-s3

## How
It updated the spec to show the option that leverages cdk logic to handle the new record type and add `file_size()` and `get_file()` to the stream reader.

## Review guide

1. `airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py`:  file_size() and get_file() to the stream reader
2. `airbyte-integrations/connectors/source-s3/source_s3/v4/config.py`: update spec.
3. `docs/integrations/sources/s3.md`: updated doc.

Note: We need to merge first this [other PR.](https://github.com/airbytehq/airbyte/pull/48345)

## User Impact
If they just need to move files as they are, this will greatly improve users' experience as sync will be quite fast.

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Review by aaronsteers — APPROVED** [review]
No surprises - all looks good! 👍 

🚀

File: airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py [hunk 1]
```diff
@@ -183,6 +187,40 @@ def open_file(self, file: RemoteFile, mode: FileReadMode, encoding: Optional[str
         # we can simply return the result here as it is a context manager itself that will release all resources
         return result
 
+    @override
+    def get_file(self, file: RemoteFile, local_directory: str, logger: logging.Logger):
```
[the remark this answers is not in the document]
Comment (aldogonzalez8) [reply airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py:191]: Makes sense, let me do that. I will add it to the CDK side as well.

File: airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py [hunk 2]
```diff
@@ -183,6 +187,40 @@ def open_file(self, file: RemoteFile, mode: FileReadMode, encoding: Optional[str
...
+        absolute_file_path = path.abspath(local_file_path)
+        logger.info(
+            f"Starting to download the file {file.uri} with size: {file_size / (1024 * 1024):,.2f} MB ({file_size / (1024 * 1024 * 1024):.2f} GB)"
+        )
+        # at some moment maybe we will require to play with the max_pool_connections and max_concurrency of s3 config
+        start_download_time = time.time()
+        self.s3_client.download_file(self.config.bucket, file.uri, local_file_path)
+        write_duration = time.time() - start_download_time
+        logger.info(f"Finished downloading the file {file.uri} and saved to {local_file_path} in {write_duration:,.2f} seconds.")
+        return {"file_url": absolute_file_path, "bytes": file_size, "file_relative_path": file_relative_path}
```
Comment (aaronsteers) [inline airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py:214]: Do we also share somewhere the original full URL?

Like `s3://my_bucket/path/to/file`?

Comment (aldogonzalez8) [reply airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py:214]: Do you mean for the record? We do it here in the CDK [here](https://github.com/airbytehq/airbyte/blob/497be1c2857615f4e3cab86e9ee09c451063eda2/airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py#L104).

**Pull request merged**